src/components/swap/ConfirmSwapModal.tsx

```diff
@@ -1,39 +1,170 @@
 import SwapModalFooter from './SwapModalFooter'
 import SwapModalHeader from './SwapModalHeader'
 
+export enum ConfirmModalState {
+  REVIEWING,
+  APPROVING_TOKEN,
+  PERMITTING,
+  PENDING_CONFIRMATION,
+}
+
+function isInApprovalPhase(confirmModalState: ConfirmModalState) {
+  return confirmModalState === ConfirmModalState.APPROVING_TOKEN || confirmModalState === ConfirmModalState.PERMITTING
+}
+
+function useConfirmModalState({
+  trade,
+  allowedSlippage,
+  onSwap,
+  allowance,
+  doesTradeDiffer,
+}: {
+  trade: InterfaceTrade
+  allowedSlippage: Percent
+  onSwap: () => void
+  allowance: Allowance
+  doesTradeDiffer: boolean
+}) {
+  const [confirmModalState, setConfirmModalState] = useState<ConfirmModalState>(ConfirmModalState.REVIEWING)
+  const [approvalError, setApprovalError] = useState<PendingModalError>()
+  const [pendingModalSteps, setPendingModalSteps] = useState<PendingConfirmModalState[]>([])
+
+  // This is a function instead of a memoized value because we do _not_ want it to update as the allowance changes.
+  // For example, if the user needs to complete 3 steps initially, we should always show 3 step indicators
+  // at the bottom of the modal, even after they complete steps 1 and 2.
+  const prepareSwapFlow = useCallback(() => {
+    const steps: PendingConfirmModalState[] = []
+    if (allowance.state === AllowanceState.REQUIRED && allowance.needsPermit2Approval) {
+      steps.push(ConfirmModalState.APPROVING_TOKEN)
+    }
+    if (allowance.state === AllowanceState.REQUIRED && allowance.needsSignature) {
+      steps.push(ConfirmModalState.PERMITTING)
+    }
+    steps.push(ConfirmModalState.PENDING_CONFIRMATION)
+    setPendingModalSteps(steps)
+  }, [allowance])
+
+  const { chainId } = useWeb3React()
+  const trace = useTrace()
+  const maximumAmountIn = useMaxAmountIn(trade, allowedSlippage)
+
+  const startSwapFlow = useCallback(async () => {
+    setApprovalError(undefined)
+    if (allowance.state === AllowanceState.REQUIRED) {
+      // Starts the approval process, by triggering either the Token Approval or the Permit signature.
+      try {
+        if (allowance.needsPermit2Approval) {
+          setConfirmModalState(ConfirmModalState.APPROVING_TOKEN)
+          await allowance.approve()
+          sendAnalyticsEvent(InterfaceEventName.APPROVE_TOKEN_TXN_SUBMITTED, {
+            chain_id: chainId,
+            token_symbol: maximumAmountIn?.currency.symbol,
+            token_address: maximumAmountIn?.currency.address,
+            ...trace,
+          })
+        } else {
+          setConfirmModalState(ConfirmModalState.PERMITTING)
+          await allowance.permit()
+        }
+      } catch (e) {
+        setConfirmModalState(ConfirmModalState.REVIEWING)
+        if (didUserReject(e)) {
+          return
+        }
+        console.error(e)
+        setApprovalError(
+          allowance.needsPermit2Approval ? PendingModalError.TOKEN_APPROVAL_ERROR : PendingModalError.PERMIT_ERROR
```

Comment:
it would be really great in the future to break up the `allowance` hook into two separate ones because token approval and permitting are technically two different things and there's no reason to bundle them together

Reply:
https://linear.app/uniswap/issue/WEB-2079/[permit2-follow-up]-split-up-the-allowance-hook